includes/class-wc-payments-apple-pay-registration.php

```diff
@@ -0,0 +1,357 @@
+		) {
+			return;
+		}
+
+		$path = WCPAY_ABSPATH . '/' . self::DOMAIN_ASSOCIATION_FILE_NAME;
+		header( 'Content-Type: text/plain;charset=utf-8' );
+		echo esc_html( @file_get_contents( $path ) ); // @codingStandardsIgnoreLine
+		exit;
+	}
+
+	/**
+	 * Processes the Apple Pay domain verification.
+	 *
+	 * @return bool Whether domain verification succeeded.
+	 */
+	public function register_domain_with_apple() {
+		$error = null;
+		try {
+			$registration_response = $this->payments_api_client->register_domain_with_apple( $this->domain_name );
+
+			if ( isset( $registration_response['id'] ) ) {
+				$this->gateway_settings['apple_pay_verified_domain'] = $this->domain_name;
+				$this->gateway_settings['apple_pay_domain_set']      = 'yes';
+				update_option( 'woocommerce_woocommerce_payments_settings', $this->gateway_settings );
+
+				Logger::log( __( 'Your domain has been verified with Apple Pay!', 'woocommerce-payments' ) );
+
+				return true;
+			} elseif ( isset( $registration_response['error']['message'] ) ) {
+				$error = $registration_response['error']['message'];
+			}
+		} catch ( Exception $e ) {
+			$error = $e->getMessage();
+		}
+
+		// Display error message in notice.
+		$this->apple_pay_verify_notice = $error;
+
+		$this->gateway_settings['apple_pay_verified_domain'] = $this->domain_name;
+		$this->gateway_settings['apple_pay_domain_set']      = 'no';
+		update_option( 'woocommerce_woocommerce_payments_settings', $this->gateway_settings );
+
+		Logger::log( 'Error: ' . $error );
+
+		return false;
+	}
+
+	/**
+	 * Process the Apple Pay domain verification if proper settings are configured.
+	 */
+	public function verify_domain_if_configured() {
+		if ( ! $this->is_enabled() ) {
+			return;
+		}
+
+		// Ensure that domain association file will be served.
+		flush_rewrite_rules();
+
+		// The rewrite rule method doesn't work if permalinks are set to Plain.
+		// Create/update domain association file by copying it from the plugin folder as a fallback.
+		$this->update_domain_association_file();
+
+		// Register the domain with Apple Pay.
+		$verification_complete = $this->register_domain_with_apple();
+	}
+
+	/**
+	 * Conditionally process the Apple Pay domain verification after settings are initially set.
+	 *
+	 * @param string $option   Option name.
+	 * @param array  $settings Settings array.
+	 */
+	public function verify_domain_on_new_settings( $option, $settings ) {
+		$this->verify_domain_on_updated_settings( [], $settings );
+	}
+
+	/**
+	 * Conditionally process the Apple Pay domain verification after settings are updated.
+	 *
+	 * @param array $prev_settings Settings before update.
+	 * @param array $settings      Settings after update.
+	 */
+	public function verify_domain_on_updated_settings( $prev_settings, $settings ) {
+		// Grab previous state and then update cached settings.
+		$this->gateway_settings = $prev_settings;
+		$prev_is_enabled        = $this->is_enabled();
```

Comment:
This is pretty interesting to read. Do you think we could make this more graceful somehow?

It's a bit difficult to see what's going on here because a reader might not know that `$this->is_enabled()` uses values from `$this->gateway_settings`. So if someone comes back to this they might remove the line updating the gateway settings before checking `is_enabled()` causing a pretty big bug.

The first improvement that comes to mind is to have some private function that checks settings passed to it, like `is_apple_pay_enabled_given_settings( array $some_settings )`. We can call the private method here with `$prev_settings`, and the public facing `is_enabled()` can call that function with `$this->gateway_settings`.

I hope that made sense 😅 

I'm also open to other improvements here. It's difficult to see precisely what's going on with these 2 lines of code and how they're connected.

Reply:
I agree this is confusing, I remember I almost deleted that same line after a first glance.

I've refactored the `is_enabled` function to support an optional `$settings` parameter, similar to what you suggested, but taking advantage of the same function. Let me know what you think.

Reply:
Looks good!